lrmd/lrmd.c

```diff
@@ -160,6 +164,96 @@ build_rsc_from_xml(xmlNode * msg)
     return rsc;
 }
 
+static void
+dup_attr(gpointer key, gpointer value, gpointer user_data)
+{
+    g_hash_table_replace(user_data, strdup(key), strdup(value));
+}
+
+static gboolean
+valid_version_format(const char *version)
+{
+    if (version == NULL) {
+        return FALSE;
+    }
+
+    if (version_format_regex == NULL) {
+        const char *regex_string = "^[[:digit:]]+([.][[:digit:]]+)*$";
+        version_format_regex = calloc(1, sizeof(regex_t));
+        regcomp(version_format_regex, regex_string, REG_EXTENDED);
+    }
+
+    return regexec(version_format_regex, version, 0, NULL, 0) == 0;
+}
+
+static const char*
+get_ra_version(const char *class, const char *provider, const char *type, gboolean update_hash)
+{
+    int len = 0;
+    char *key = NULL;
+    const char *default_version = "0.1";
+    const char *version = NULL;
+
+    CRM_CHECK(type != NULL, return default_version);
+    CRM_CHECK(class != NULL, return default_version);
+
+    if (!provider) {
+        provider = "heartbeat";
+    }
+    
+    len = strlen(class) + strlen(provider) + strlen(type) + 3;
+    key = calloc(len, sizeof(char));
+
+    if (!key) {
+        return default_version;
+    }
+    
+    if (!ra_version_hash) {
+        ra_version_hash = g_hash_table_new_full(crm_str_hash, g_str_equal, g_hash_destroy_str, g_hash_destroy_str);
+    }
+
+    snprintf(key, len, "%s:%s:%s", class, provider, type);
+
+    version = g_hash_table_lookup(ra_version_hash, key);
+
+    if (!version || update_hash) {
+        char *metadata = NULL;
+        lrmd_t *lrmd = lrmd_api_new();
+
+        lrmd->cmds->get_metadata(lrmd, class, provider, type, &metadata, 0);
+        lrmd_api_delete(lrmd);
+
+        if (metadata) {
+            xmlNode *metadata_xml = string2xml(metadata);
+            xmlNode *version_xml = get_xpath_object("//resource-agent/@version", metadata_xml, LOG_TRACE);
+
+            if (version_xml) {
+                version = crm_element_value(version_xml, XML_ATTR_VERSION);
```

Comment:
Well I guess my mistake was that I haven't strdupped key and value before inserting them into ra_version_hash. Since we now only using get_ra_version() in lrmd and this hash table lives as long as the process do, I think it will be sufficient to return const char* pointing into version contained in it.

I also changed the default_version pointer to static. Thanks for figuring this out for me.